### ISSUE TITLE
Chore: Move UI-related C++ components to a common location

### ISSUE THREAD

**Issue by micieslak** [status-im/status-desktop#9125]
## Description

Currently C++ components are located in `status-desktop/vendor/DOtherSide/lib`. It's problematic because some C++ based UI components would be useful also in StatusQ (e.g. `RXValidator` type). Not using C++ types seems to be a big limitation here. Additionally components like `Storybook` (`status-desktop/storybook`) also use some C++ types. Currently sharing those C++ components between Storybook and other parts of the app is problematic. E.g. it would be useful to share C++ components for hot reloading between Storybook and Monitoring app.

Currently there is also other minor issues with sharing C++ code - `SortFilterProxyModel` is a submodule in both DOtherSide (`status-desktop/vendor/DOtherSide/vendor/SortFilterProxyModel`) and in StatusQ (`status-desktop/ui/StatusQ/vendor/SortFilterProxyModel`).

The proposed solution is to keep C++ types in their "natural" place, and refer to them from DOtherSide configuration:
- things like `RXValidator` could be kept directly in `StatusQ` because it's logically part of StatusQ (generic UI component)
- `SortFilterProxyModel` could be a submodule only in StatusQ and referred from DOtherSide config as well
- Storybook/Monitoring c++ types could be moved to StatusQ as well bc they are generic, or alternatively kept in shared location in the main project directory.

The drawback of that solution is fact, that then keeping `DOtherSide` as a separate module doesn't make much sense then. It would refer to locations specific for the project outside of it's own directory tree.

**Comment by caybro** [comment]
Yes, very much agree

**Comment by alexandraB99** [comment]
+1 from me and also to consider `SyntaxHighligther`

**Comment by micieslak** [comment]
> +1 from me and also to consider `SyntaxHighligther`

Yeah, everything like that. So do we agree on merging `DOtherSide` into `status-desktop` like we did (but for other reason) with `StatusQ`?

**Comment by caybro** [comment]
imho you don't necessarily need to separate the stuff in the first step; merely having DOtherSide in one place (and built only once!) is already a considerable improvement

**Comment by osmaczko** [comment]
Well, tbh I don't like the idea of making `DOtherSide` dependency bidirectional. It's generally better to keep dependencies unidirectional and well-defined, especially that there is also `nimqml` in between.

> Currently C++ components are located in `status-desktop/vendor/DOtherSide/lib`. It's problematic because some C++ based UI components would be useful also in StatusQ (e.g. `RXValidator` type). Not using C++ types seems to be a big limitation here.

Please note that `StatusQ` is not a separate library, it is only logically a separate part of our app and we can use all the utilities exposed from `DOtherSide` there, same as status-app:
![image](https://user-images.githubusercontent.com/33099791/212711612-3beacdbf-8876-4f6b-9998-865329ea907c.png)

> Currently there is also other minor issues with sharing C++ code - SortFilterProxyModel is a submodule in both DOtherSide (status-desktop/vendor/DOtherSide/vendor/SortFilterProxyModel) and in StatusQ (status-desktop/ui/StatusQ/vendor/SortFilterProxyModel).

That's only because of `sanbox-app`. It can be easily fixed by linking to `DOtherSide`, it will also make `sanbox-app` work if StatusQ use types registered in `DOtherSide`:

![image](https://user-images.githubusercontent.com/33099791/212712674-286f75c1-7cd6-4716-9504-20bd7f6debcb.png)

**Comment by osmaczko** [comment]
Well, tbh, I don't like the idea of making `DOtherSide` a bidirectional dependency. It's generally better to keep dependencies unidirectional and well-defined, especially that there is `nimqml` in between.

> Currently C++ components are located in `status-desktop/vendor/DOtherSide/lib`. It's problematic because some C++ based UI components would be useful also in StatusQ (e.g. `RXValidator` type). Not using C++ types seems to be a big limitation here.

`StatusQ` is not a separate library, it is just a logical part of our app, nothing else. It means, we can use everything exposed through `DOtherSide` the same as we do for `nim-app`:
![image](https://user-images.githubusercontent.com/33099791/212717895-8cf3726d-ee80-44ea-920f-009021b87578.png)

> Currently there is also other minor issues with sharing C++ code - SortFilterProxyModel is a submodule in both DOtherSide (status-desktop/vendor/DOtherSide/vendor/SortFilterProxyModel) and in StatusQ (status-desktop/ui/StatusQ/vendor/SortFilterProxyModel).

That's only because of `sandbox-app` (that we no longer support?). It can be easily solved by linking it against `DOtherSide` rather than `SortFilterProxyModel`. It will also make `sandbox-app` work properly if any `StatusQ` component uses something exposed by `DOtherSide`.

![image](https://user-images.githubusercontent.com/33099791/212718647-1fae90e1-6303-42f3-a697-b75ef1ba1a0f.png)

We don't probably want to link against `DoOtherSide` as it is too heavy. 

From the separation of concerns point of view, and to keep dependencies unidirectional the following would be ideal:
![image](https://user-images.githubusercontent.com/33099791/212719051-cb5b0c21-06e6-4df7-a028-ca995c18c24d.png)

We can make `status-cpp-extensions` a part of `DOtherSide` to avoid the burden of creating PRs in many different repos.

**Comment by micieslak** [comment]
Thank you @osmaczko for your response. It turns out that we have different understanding about that part. Let's clarify it a bit to find a golden mean :)

> StatusQ is not a separate library, it is just a logical part of our app, nothing else.

From what I understand, we want to keep it as was a separate library. It was in a separate repo but we merged it only for convenience. And we created guards to keep dependency unidirectional and library-like form. It's not built as a library, that's true, but it's quite self-contained, has own demo-app and tests, configured in a standard way via cmake.

> It means, we can use everything exposed through DOtherSide the same as we do for nim-app

How we could do that? `Sandbox`'s cmake config would have to refer to `DOtherSide` location in `status-desktop`. This location is of course outside of `StatusQ` base dir. It would make (currently totally independent) `StatusQ` heavily dependent on stuff that we keep in specific locations of `status-desktop`.

The approach from the third diagram imo has the same problem. It would work only if we would stop treating StatusQ as a self-contained lib and exclude sandbox-app to upper directory (to avoid referring like `../../vendor/DOtherSide`).

The idea behind this task is to keep c++ stuff in proper place, not artificial place enforced by nim-related limitations:

- cpp ui componets, like e.g. `RxValidator` would be kept within `StatusQ` directory. It could be used in other, higher level qml components in a regular way. For `sandbox-app` app or `tests` config it would be easily accessible (no need to refer external content).
- cpp components for Storybook could be similarly located in `storybook` directory
- cpp components specific for `status-desktop` (not sure if we have them right now) could be stored also in "standard" location for them, not in a very speficic external submodule.

In this approach role of `DOtherSide` should be understood in a slightly different way than so far.
It should be treated just as a tool allowing building the qt/qml app combined with some nim sources. It just takes c++ sources needed by the app (e.g. from StatusQ or direct c++ components of status-desktop) and builds them because while `nim` is involved it cannot be done in a standard way (regular cmake config). In this approach `DOtherSide` doesn't contain any sources which logically belong to other modules (status-desktop, StatusQ, Storybook). So imo there is no bidirectional dependency imo.

Rephrasing:
- StatusQ uses standard cmake-based config for building (sandbox, tests). C++ sources are just within `StatusQ`'s dir.
- Storybook also uses standard cmake-based config. Specific for Storybook c++ components (e.g. custom cpp models) are also stored in Storybook's location
- status-desktop doesn't use standard cmake config because of Nim. Therefore it uses `DOtherSide` to accomplish this goal. Specific c++ components for status-desktop are located in a standard location (more or less like in other qt/qml projects). `DOtherSide` as a build helper refers to all those locations needed to build the app and create lib cooperating smoothly with rest of nim logic and qml part.

**Comment by osmaczko** [comment]
> From what I understand, we want to keep it as was a separate library. It was in a separate repo but we merged it only for convenience. And we created guards to keep dependency unidirectional and library-like form. It's not built as a library, that's true, but it's quite self-contained, has own demo-app and tests, configured in a standard way via cmake.

If you take a look at it that way, then yes, StatusQ is a separate package. For nim-app StatusQ is just a qml subdirectory, it is not built as a library and the sources are treated as if they belong to nim-app. That's why when you do `import DotherSide 0.1` in any StatusQ component and run the nim-app, it will simply work. Of course, it would break sandbox-app and imports checker, to fix that we would need to make StatusQ DOtherSide aware, as mentioned before.

> How we could do that? Sandbox's cmake config would have to refer to DOtherSide location in status-desktop. This location is of course outside of StatusQ base dir. It would make (currently totally independent) StatusQ heavily dependent on stuff that we keep in specific locations of status-desktop.

Well, we either keep StatusQ totally independent and provide the DOtherSide the same way as we do with SortFilterProxyModel now, effectively duplicating nim-app repo dependencies or we do a compromise and refer to the parent directory for vendors. I wouldn't name it a "heavy" dependency.

> The idea behind this task is to keep c++ stuff in proper place, not artificial place enforced by nim-related limitations

Yes, that makes a lot of sense to me. On one hand, making DoOtherSide dependent on StatusQ screams "no, please don't", on the other, bringing the benefits you mention makes it feel ok.

> So imo there is no bidirectional dependency imo.

Agreed, there is no bidirectional dependency.
![image](https://user-images.githubusercontent.com/33099791/212871966-eda19b2e-b26a-4132-9ab9-14a3aac04ef9.png)

All in all, your arguments convinced me, green light from me.
